src/rgw/rgw_bucket.cc

```diff
@@ -1088,24 +947,130 @@ int RGWBucketAdminOp::dump_s3_policy(rgw::sal::Store* store, RGWBucketAdminOpSta
 
 int RGWBucketAdminOp::unlink(rgw::sal::Store* store, RGWBucketAdminOpState& op_state, const DoutPrefixProvider *dpp)
 {
-  RGWBucket bucket;
-
-  int ret = bucket.init(store, op_state, null_yield, dpp);
-  if (ret < 0)
-    return ret;
-
-  return bucket.unlink(op_state, null_yield, dpp);
+  return static_cast<rgw::sal::RadosStore*>(store)->ctl()->bucket->unlink_bucket(op_state.get_user_id(), op_state.get_bucket()->get_info().bucket, null_yield, dpp, true);
```

Comment:
shouldn't these be dynamic_casts with nullptr checks?

Reply:
All these casts are temporary, and will be removed in the refactor PR.  The goal is to have no more casts at all.